### PR TITLE
Fix issue 19572: Prevent Appender from implicitly calling struct constructors to cast away const.

### PR DESCRIPTION
**Pull request by FeepingCreature**
Prevent Appender from implicitly calling struct constructors to cast away const.

Sample code:

```
struct Struct
{
    int value;

    int fun() const { return 23; }

    alias fun this;
}

void main()
{
    import std.array : Appender;

    Appender!(Struct[]) appender;

    appender.put(const(Struct)(42));

    auto result = appender.data[0];

    assert(result.value != 23);
}
```

What happens is that `Appender` checks that it can convert `const(Struct)` implicitly to `Struct`, but it does so via `cast(Struct)`, which then invokes `Struct(const(Struct).fun)` implicitly.

With this fix, instead of checking `const(Struct)` vs `Struct`, we check `Unqual` against `Unqual` and then use the dmd-fixed `cast()` to remove constness, before relying on implicit conversion to do the rest.

Note that this relies on https://github.com/dlang/dmd/pull/9223 and stems from the same investigation.

### CODE REVIEW

**Comment by FeepingCreature** [comment]
To emphasize: it is expected that this build fails while the DMD fix is not yet merged.

File: std/array.d [hunk 1]
```diff
@@ -3657,6 +3657,26 @@ if (isDynamicArray!A)
...
+        alias fun this;
+    }
+
+    Appender!(Struct[]) appender;
+
+    appender.put(const(Struct)(42));
+
+    auto result = appender.data[0];
+
+    assert(result.value != 23);
```
Comment (thewilsonator) [inline std/array.d]: Apparently this fails.

Comment (FeepingCreature) [reply std/array.d]: Correct, since it depends on the ongoing DMD PR.

Comment (thewilsonator) [reply std/array.d]: Oh whoops. That needs a changelog but otherwise looks good.

Comment (wilzbach) [reply std/array.d]: Changelog will be automatically generated as it mentions an issue

Comment (thewilsonator) [reply std/array.d]: No, without a changelog it gets a generic `issue NNNNN was fixed` entry. A specific changelog entry is for highlighting things that may be of interest, e.g. w.r.t changing behaviour such as this.

Comment (wilzbach) [reply std/array.d]: (I was talking about this specific PR here which doesn't need a detailed changelog.)

**Comment by FeepingCreature** [comment]
DMD fix merged, seems green now.

**Pull request merged**